File: js/problem_4.js
/*
A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.

Find the largest palindrome made from the product of two 3-digit numbers.
*/

// solution has correct answer in output... is dirty.
//TODO- Refactor w/ PE solution spec
function isPalindrome(num){
  let str = num.toString();
  let n = str.length / 2;

  if (str.slice(0,n) == str.slice(-n).split("").reverse().join("")) {
    return num;
  }
  return false;
}

function largestPalindrome(){
  for (let i = 99; i <= 1000; i++){
    for (let j = i + 1; j <= 1000; j++){
      let product = i * j;
      if (isPalindrome(product))  {
        console.log(i, j, product);
      }
    }
  }
}
console.log(largestPalindrome());

File: c/problem_4.c
#include <stdio.h>
#include <stdbool.h>

bool isPalindrome(int n) {
  int np = 0; 
  for (int r = n; r > 0; r /= 10) {
    np *= 10; 
    np += r % 10; 
  }
  return n == np; 
}

int main(void){
  int largestPalindrome = 0; 
  int max = 999; 
  for (int a = max; a >= 0; --a) {
    for (int b = max; b >= 0; --b) {
      if (isPalindrome(a*b) && a * b > largestPalindrome) {
        largestPalindrome = a * b; 
      }
    }
  }
  printf("%d\n", largestPalindrome); 
  return 0; 
}

